test/serviceprofiles/serviceprofiles_test.go

```diff
@@ -117,7 +135,153 @@ func TestServiceProfiles(t *testing.T) {
+
+			out, stderr, err = TestHelper.LinkerdRun(cmd...)
+			if err != nil {
+				t.Errorf("'linkerd %s' command failed with %s: %s\n", cmd, err.Error(), stderr)
+			}
+
+			_, err = TestHelper.KubectlApply(out, testNamespace)
+			if err != nil {
+				t.Errorf("kubectl apply command failed:\n%s", err)
+			}
+
+			assertion := &routeStatAssertion{
+				upstream:   testUpstreamDeploy,
+				downstream: testDownstreamDeploy,
+				namespace:  testNamespace,
+			}
+			switch tc {
+			case "retries":
+				// If the effective success rate is not equal to the actual success rate retries might already
+				// be applied so we fail the test.
+				assertion.assertFunc = func(rt *rowStat) bool { return rt.EffectiveSuccess == rt.ActualSuccess }
+				assertion.routeProperty = "Effective Success"
+				assertion.expected = "Effective Success == Actual Success"
+			case "timeouts", "budgets":
+				// If the P99 latency is greater than 500ms retries are probably happening before applying
+				// the service profile and we can't reliably test the service profile.
+				assertion.assertFunc = func(rt *rowStat) bool { return rt.LatencyP99 < 500 }
+				assertion.routeProperty = "P99 Latency"
+				assertion.expected = "< 500ms"
+			}
+
+			assertRouteStat(assertion, t)
+
+			profile := &sp.ServiceProfile{}
+
+			// Grab the output and convert it to a service profile object for modification
+			err = yaml.Unmarshal([]byte(out), profile)
+			if err != nil {
+				t.Errorf("unable to unmarshall YAML: %s", err.Error())
+			}
+
+			for _, route := range profile.Spec.Routes {
+				if route.Name == "GET /testpath" {
+					route.IsRetryable = true
+					route.Timeout = "500ms"
+
+					if tc == "budgets" {
+						profile.Spec.RetryBudget = &sp.RetryBudget{
+							RetryRatio:          1.0,
+							MinRetriesPerSecond: 10,
+							TTL:                 "10s",
+						}
+					}
+					break
+				}
+			}
+
+			bytes, err := yaml.Marshal(profile)
+			if err != nil {
+				t.Errorf("error marshalling service profile: %s", bytes)
+			}
+
+			out, err = TestHelper.KubectlApply(string(bytes), testNamespace)
+			if err != nil {
+				t.Errorf("kubectl apply command failed:\n%s :%s", err, out)
+			}
+
+			switch tc {
+			case "retries":
+				// If we get an effective success rate of less than or equal to the actual success rate requests are not
+				// being retried successfully after we applied our modified service profile.
+				assertion.assertFunc = func(rt *rowStat) bool { return rt.EffectiveSuccess > rt.ActualSuccess }
+				assertion.routeProperty = "Effective Success"
+				assertion.expected = "> Actual Success"
+				// If we get a P99 latency of less than 250ms then we aren't hitting the timeout limit
+				// set in in service profile. hello-timeouts-service and hello-budgets always fails
+				// so we expect all request latencies to be greater than or equal to the timeout set.
+			case "timeouts", "budgets":
+				assertion.assertFunc = func(rt *rowStat) bool { return rt.LatencyP99 >= 500 }
```

Comment:
From my understanding of the initial ticket, I thought we were trying to specifically test the `timeouts` behavior. Maybe, since its unreliable we should just test that retries are being performed?

Reply:
> Maybe, since its unreliable we should just test that retries are being performed?

SGTM, unless others have a way to do this. It isn't easy to reliably test timeouts without [`context`](https://golang.org/pkg/context/). So far, I can't find anything in the service profile and route code that will be useful here, based on the little that I know.